cluster/src/main/scala/org/apache/spark/scheduler/SnappyTaskSchedulerImpl.scala

```diff
@@ -16,13 +16,259 @@
+          //    In this case, send a kill signal to the executors to kill the task
+          //    and then abort the stage.
+          // 2. The task set manager has been created but no tasks has been scheduled.
+          //    In this case, simply abort the stage.
+          tsm.runningTasksSet.foreach { tid =>
+            val execId = taskIdExecutorAndManager.get(tid)._1
+            backend.killTask(tid, execId, interruptThread)
+          }
+          val msg = s"Stage $stageId cancelled"
+          tsm.abort(msg)
+          logInfo(msg)
+          true
+        }
+      })
+    }
+  }
+
+  override def taskSetFinished(manager: TaskSetManager): Unit = synchronized {
+    val taskSet = manager.taskSet
+    stageCoresAndAttempts.get(taskSet.stageId) match {
+      case null =>
+      case (_, taskSetsForStage) =>
+        taskSetsForStage.remove(taskSet.stageAttemptId)
+        if (taskSetsForStage.size() == 0) {
+          stageCoresAndAttempts.remove(taskSet.stageId)
+        }
+    }
+    manager.parent.removeSchedulable(manager)
+    if (isInfoEnabled) {
+      logInfo(s"Removed TaskSet ${taskSet.id}, whose tasks have all completed, from pool " +
+          manager.parent.name)
+    }
+  }
+
+  /**
+   * Avoid giving all the available cores on a node to a single task. This serves two purposes:
+   *
+   * a) Keeps some cores free for any concurrent tasks that may be submitted after
+   * the first has been scheduled.
+   *
+   * b) Since the snappy executors use (2 * physical cores) to aid in more concurrency,
+   * it helps reduce disk activity for a single task and improves performance for
+   * disk intensive queries.
+   */
+  override protected def resourceOfferSingleTaskSet(
+      taskSet: TaskSetManager,
+      maxLocality: TaskLocality,
+      shuffledOffers: Seq[WorkerOffer],
+      availableCpus: Array[Int],
+      tasks: IndexedSeq[ArrayBuffer[TaskDescription]]): Boolean = {
+    // reduce the available CPUs for a single taskSet if more than physical cores are exposed
+    val availableCores = if (numExecutors.get() > 0) {
+      val coresAndAttempts = stageCoresAndAttempts.get(taskSet.taskSet.stageId)
+      if (coresAndAttempts ne null) coresAndAttempts._1 else null
+    } else null
+    var launchedTask = false
+    for (i <- shuffledOffers.indices) {
+      val execId = shuffledOffers(i).executorId
+      if ((availableCpus(i) >= CPUS_PER_TASK) &&
+          ((availableCores eq null) ||
+              (availableCores.computeIfAbsent(execId, lookupExecutorCores) >= CPUS_PER_TASK))) {
+        try {
+          val host = shuffledOffers(i).host
+          for (task <- taskSet.resourceOffer(execId, host, maxLocality)) {
+            tasks(i) += task
+            val tid = task.taskId
+            taskIdExecutorAndManager.justPut(tid, execId -> taskSet)
+            executorIdToRunningTaskIds(execId).add(tid)
+            if (availableCores ne null) {
+              availableCores.addValue(execId, -CPUS_PER_TASK)
```

Comment:
Can we put an assertion similar to  assert(availableCpus(i) >= 0) ?
We might catch some of the erroneous updates.

Reply:
Will add.